src/main/java/org/bytedeco/javacpp/tools/Generator.java

```diff
@@ -937,23 +938,44 @@ boolean classes(boolean handleExceptions, boolean defineAdapters, boolean conver
         out.println("}");
         out.println();
         if (handleExceptions || convertStrings) {
-            out.println("static JavaCPP_noinline jstring JavaCPP_createString(JNIEnv* env, const char* ptr) {");
+            out.println("#include <string>");
+            out.println("static JavaCPP_noinline jstring JavaCPP_createStringUTF8(JNIEnv* env, const char* ptr, size_t length) {");
```

Comment:
Methods like `String(byte[])` and `String.getBytes()` actually work for the native charset of the underlying OS, not necessarily UTF-8. We could add custom methods that somehow work with a user-specified encoding, but the default behavior of those methods should stay the same, and they shouldn't be named "UTF8".

Reply:
Yeah, I missed that.
I actually started to implement a more generic solution that would allow to pass any Charset to createString/getBytes (so that String could be created from byte array in any encoding), when I thought: how do we determine that byte array returned from getBytes() can be safely null-terminated?. I mean, when we are using default charset, there is no guarantee that returned array won't have null byte somewhere in the middle. Do we just shift responsibility to user?

Reply:
The null character gets added here manually after GetByteArrayRegion(), just like you did with GetStringRegion():
https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Generator.java#L972
As for passing a Charset, I'm thinking something defined globally at compile time a per issue https://github.com/bytedeco/javacpp/issues/70#issuecomment-211138577 would be best. I'm not sure how we'd go about passing a Java *object* rather than just a string like that though. The latter entails potential exceptions at initialization time, but I guess that wouldn't so bad. What do you think?

Reply:
I was thinking about passing a charset name to `getBytes(java.lang.String)`/`String(byte[], java.lang.String)` overloads. Generator would extract charset name from annotation, hardcode it in C++ code which will create jstring from it using NewStringUTF (we can use modified utf-8 here I believe) and pass it to to getString()/String(). It has an advantage over define in that you could use different charsets for different functions.

The downside is that these NewStringUTF() calls will introduce additional overhead (JMH said that it would be almost 50% slower than using `String(byte[])`) unless you find some way to cache them.

Setting encoding via define would of course also require passing jstring with charset name (unless you meant some other method?) but then you can easily cache it.

Reply:
I'm thinking something like this too, but the issue with passing a charset name like that for every call is that it may through an exception at runtime, any time. It's better to resolve it at load time into a Charset object.